### PR TITLE
Expand understanding of build targets in et

### PR DESCRIPTION
**Pull request by johnmccutchan**
- s/TestTarget/BuildTarget.
- Use a more informative way of querying for build targets from gn
- Port existing code to use new interfaces
- Replace 'query tests' with 'query targets' and a --testonly flag
- Extend 'et build' with support for build target selectors.
- Extend 'et query targets' with support for build target selectors.

### CODE REVIEW

File: tools/engine_tool/lib/src/commands/test_command.dart [hunk 1]
```diff
@@ -50,19 +50,29 @@ final class TestCommand extends CommandBase {
...
-    final Map<String, TestTarget> allTargets = await findTestTargets(
-        environment,
+    final Map<String, BuildTarget> allTargets = await findTargets(environment,
         Directory(p.join(environment.engine.outDir.path, build.ninja.config)));
-    final Set<TestTarget> selectedTargets =
+    final Set<BuildTarget> selectedTargets =
         selectTargets(argResults!.rest, allTargets);
     if (selectedTargets.isEmpty) {
       environment.logger.error(
           'No build targets matched ${argResults!.rest}\nRun `et query tests` to see list of targets.');
```
Comment (matanlurey) [inline tools/engine_tool/lib/src/commands/test_command.dart:59]: Is et query tests still right? A test only _target_ is not the same as a test (I think that's what you're fixing?)

File: tools/engine_tool/lib/src/commands/query_command.dart [hunk 2]
```diff
@@ -152,24 +160,38 @@ final class QueryTestsCommand extends CommandBase {
...
 
   @override
-  String get name => 'tests';
+  String get name => 'targets';
 
   @override
-  String get description => 'Provides information about test targets';
+  String get description => 'Provides information about build targets'
+      'et query targets --testonly         # List only test targets'
+      'et query targets //flutter/fml/...  # List all targets under `//flutter/fml`';
```
[the remark this answers is not in the document]
Comment (johnmccutchan) [reply tools/engine_tool/lib/src/commands/query_command.dart:168]: I think the current help text is correct:

query targets is a general command now that lists all build targets
the --testonly flag restricts the listed build targets to those marked with "testonly = 1" in BUILD.gn

File: tools/engine_tool/lib/src/commands/query_command.dart [hunk 3]
```diff
@@ -152,24 +160,38 @@ final class QueryTestsCommand extends CommandBase {
   late final List<Build> builds;
 
   @override
-  String get name => 'tests';
+  String get name => 'targets';
 
   @override
-  String get description => 'Provides information about test targets';
+  String get description => 'Provides information about build targets'
```
[the remark this answers is not in the document]
Comment (johnmccutchan) [reply tools/engine_tool/lib/src/commands/query_command.dart:166]: This command now queries all build targets not just tests now.

File: tools/engine_tool/lib/src/commands/test_command.dart (hunk 1 above)
Comment (johnmccutchan) [reply tools/engine_tool/lib/src/commands/test_command.dart:59]: This text has now been updated to call out 'et query targets'

File: tools/engine_tool/lib/src/json_utils.dart [hunk 4]
```diff
@@ -84,3 +84,42 @@ const JsonEncoder _jsonEncoder = JsonEncoder.withIndent('    ');
 
 /// Same as [jsonEncode] but is formatted to be human readable.
 String jsonEncodePretty(Object? object) => _jsonEncoder.convert(object);
+
+/// Returns the value in map[field] iff it is a String. null otherwise.
+String? getString(Map<String, Object?> map, String field) {
+  if (map[field] == null) {
```
[the remark this answers is not in the document]
Comment (johnmccutchan) [reply tools/engine_tool/lib/src/json_utils.dart:90]: Whoa.

[the remark this answers is not in the document]
Comment (johnmccutchan) [reply tools/engine_tool/lib/src/json_utils.dart:90]: Done.

File: tools/engine_tool/lib/src/commands/query_command.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (matanlurey) [reply tools/engine_tool/lib/src/commands/query_command.dart:168]: Ah ack, resolved.

**Review by cbracken — APPROVED** [review]

**Review by cbracken — APPROVED** [review]
LGTM in principle, but presubmit is failing.

Animated GIF of a delivery truck approaching a highway overpass, then failing to meet the height clearance requirement and crashing into it

File: tools/engine_tool/test/query_command_test.dart [hunk 5]
```diff
@@ -162,7 +152,7 @@ fml_arc_unittests
     );
     final int result = await runner.run(<String>[
       'query',
-      'tests',
+      'targets',
     ]);
     expect(result, equals(0));
```
[the remark this answers is not in the document]
Comment (johnmccutchan) [reply tools/engine_tool/test/query_command_test.dart:157]: Yes, fixed a bunch of test failures because the index of spawned processes changed.

**Comment by zanderso** [comment]
I believe it will be useful to have a way to retain the behavior of building the targets specified in the build config json files. Is there some way to work that behavior back into the `build` command?

File: tools/engine_tool/lib/src/gn_utils.dart [hunk 6]
```diff
@@ -106,9 +138,20 @@ Future<Map<String, TestTarget>> findTestTargets(
...
+Set<BuildTarget> selectTargets(
+    List<String> selectors, Map<String, BuildTarget> allTargets) {
+  final Set<BuildTarget> selected = <BuildTarget>{};
+
+  if (selectors.isEmpty) {
+    // Default to all if no selectors are specified.
+    // TODO(johnmccutchan): Reconsider this default or at least lift
+    // this logic up to the caller.
+    allTargets.values.forEach(selected.add);
+    return selected;
```
Comment (zanderso) [inline tools/engine_tool/lib/src/gn_utils.dart:152]: If you return empty set here, and don't make this an error at the call site of `selectTargets`, then the `extraNinjaTargets` parameter of the `BuildRunner` constructor will be empty, and the targets specified by the build config json will be used.

File: tools/engine_tool/lib/src/commands/build_command.dart [hunk 7]
```diff
@@ -53,6 +61,22 @@ final class BuildCommand extends CommandBase {
       if (!useRbe) '--no-rbe',
     ];
 
-    return runBuild(environment, build, extraGnArgs: extraGnArgs);
+    final Map<String, BuildTarget> allTargets = await findTargets(environment,
```
Comment (zanderso) [inline tools/engine_tool/lib/src/commands/build_command.dart:64]: This directory might not exist if GN hasn't had a chance to run yet. That is, if you have a fresh engine checkout and `out/` is empty, then `selectTargets` will fail.

**Pull request merged**